README
Website: http://firegpg.com
Authors: FireGPG Team <support@firegpg.com>

This extension is compatible with Firefox 3 and above, on Windows, MacOS and Linux platforms.

===============
Build the XPI
===============
You can build an XPI for installation by running the build.sh script located in this folder.

For development you should do the following:
  1. Unzip the entire contents of this package to somewhere,
	       e.g, c:\dev or /home/user/dev
  2. Put the full path to the folder (e.g. c:\dev\firegpg on
     Windows, /home/user/dev/firegpg on Linux) in a file named
     firegpg@firegpg.team and copy that file to
     [your profile folder]\extensions\
  3. Restart Firefox.

===============
Report a bug
===============
<<<<<<< HEAD
Please report your bugs on the bug tracker: http://firegpg.com/
=======
Please report your bugs on the bug tracker: http://bugs.getfiregpg.org
>>>>>>> d5e1a563dc2d10392a7a3f79cb4ea83d9d986406


ChangeLog
2010-06-13 FireGPG *NOT* discontinued.  Forked by one of the lead devs.  http://firegpg.com/

2010-06-07 Maximilien Cuony
    * http://blog.getfiregpg.org/2010/06/07/firegpg-discontinued/
    * 0.8 released !
	* Appling GpgAuth patch
    * Removing Gmail support
    * Updating locates

2009-11-08 Maximilien Cuony
    * 0.7.10 released again !
    * Applied patch from Nils Maier
    * 0.7.10 released !
    * Fix a problem with menu

2009-11-05 Maximilien Cuony
    * Renamed namespaces
    * Putting back FireGPG result on the right
    * New options:download_key_without_asking,hide_the_wait_window,hide_key_server_confirmation,hide_key_operation_confirmation,hide_download_menu
    * Added a warning about correct signs but with keys who aren't trusted

2009-10-25 Maximilien Cuony
    * Fixing regression with signed and encrypted mail
    * Fixing problem with pgp blocks inside pgp blocks
    * Applied patch of AtesComp to improve OWA support.
    * Default keyserver set to pgp.mit.edu
    * If you click on the wait box for keyservers, she is closed.
    * Moving down by one position the FireGPG menu in downloads
    * Fixing a problem in default options
    * Gmail support working again..
    * Adding detection of gpg for x64 windows
    * Fixing problem with el and he locales

2009-09-26 Maximilien Cuony
    * 0.7.9 released !
    * Basque, Finnish, Galician added


2009-09-20 Maximilien Cuony
    * Fix problem dues to namespacing
    * Improved error messages
    * Gmail's improvements
    * Fix gmail issues

2009-09-19 Maximilien Cuony
    * Max version set to 3.6
    * Asking twice symetrical passwords
    * Fixing dates on signatures lists
    * Button to show fingerprints
    * New item in menu to check again a page for pgp block
    * Typo in mime decoder
    * Fixing problem with presistant security label of pgp block
    * FireGPG is now fully namespaced !

2009-07-29 Maximilien Cuony
    * 0.7.8 released !
    * Fix problem with files signs
    * Fix problem with hashes with  fx3.0

2009-07-28 Maximilien Cuony
    * Adding id & lv locates
    * Fix gmail problem
    * Fix problem with foward in gmail

2009-07-27 Maximilien Cuony
    * Adding some parameters to very for DTA

2009-07-26 Maximilien Cuony
    * 0.7.7 released !
    * New icon theme
    * FS#325

2009-07-21 Maximilien Cuony
    * New API system

2009-07-20 Maximilien Cuony
    * Work on api's security
    * Upgraded a part of the doc

2009-07-19 Maximilien Cuony
    * Improving security of the api if gpg-agent is turned on

2009-07-18 Maximilien Cuony
    * Cancel api request when blank password given

2009-07-16 Maximilien Cuony
    * FS#331
    * FS#327

2009-07-15 Maximilien Cuony
    * Assistant finalised
    * FS#319
    * FS#326

2009-07-14 Maximilien Cuony
    * Escaping filenames with spaces
    * Translating assistant

2009-07-11 Maximilien Cuony
    * FS#308
    * FS#3
    * Doing some cleaning up in the code, including license..
    * Don't block interface when using keyservers

2009-07-09 Maximilien Cuony
    * Improving gmail detection
    * FS#321
    * FS#304
    * Begining download menu

2009-07-08 Maximilien Cuony
    * Assistant done, except translations {FS#41}

2009-07-04 Maximilien Cuony
    * Removing warning when erasing saved password
    * FS#305
    * FS#314
    * FS#315
    * FS#317
    * Old api disabled
    * FS#318
    * Improved inline size
    * !FS#313
    * Begining the assistant

2009-06-21 Maximilien Cuony
    * 0.7.6 released !
    * Autodecrypting multi inline encrypted mail

2009-06-20 Maximilien Cuony
    * FS#170
    * Cleaning locales : +el vi, -po sv bn nb nb-no sk
    * FS#292
    * FS#297 (http://eprint.iacr.org/2009/259.pdf)

2009-06-18 Maximilien Cuony
    * New website
    * Don't ask twice for download key
    * FS#301
    * FS#302

2009-06-05 Maximilien Cuony
    * FS#284

2009-06-04 Maximilien Cuony
    * FS#294
    * FS#295
    * FS#256
    * FS#257

2009-05-25 Maximilien Cuony
    * FS#253

2009-05-24 Maximilien Cuony
    * FS#271
    * FS#277
    * FS#279
    * FS#270
    * FS#276
    * FS#275
    * FS#252

2009-03-06 Maximilien Cuony
    * 0.7.5 packed, and uploaded to the server !

2009-03-06 Maximilien Cuony
    * Gmail working again

2009-02-15 Maximilien Cuony
    * 0.7.4 packed, and uploaded to the server !
    * Adding Romanian, Esperanto, Slovak locates

2009-02-14 Maximilien Cuony
    * Fixing gmails problem (mails not parsed)
    * Fixing gmails problem (mails sending)
    * FS#241
    * FS#245
    * FS#248 (no more wrap system on webmails)
    * Improving mail detection

2009-02-11 Maximilien Cuony
    * New gmail interface
    * Smaller editor

2009-02-10 Maximilien Cuony
    * FS#246
    * FS#240

2009-02-09 Maximilien Cuony
    * Fix problem with inline decryption
    * Fixed some problem with compose buttons

2009-02-08 Maximilien Cuony
    * 0.7.3 packed, and uploaded to the server !
    * Small bugs fixed

2009-02-07 Maximilien Cuony
    * Manage encrypted attachements and subpart

2009-02-06 Maximilien Cuony
    * Allow importing private key with the keymananger
    * New mime decoder

2009-02-05 Maximilien Cuony
    * Gmail working again
    * FS#217
    * FS#229
    * Refresh keylist when importing a file
    * FS#230

2009-02-02 Maximilien Cuony
    * Fix problem with mails sended by claws mail, kmail, Mutt

2009-02-01 Maximilien Cuony
    * FS#225
    * FS#224
    * FS#219
    * FS#226
    * FS#221

2009-01-31 Maximilien Cuony
    * FS#228
    * Splitting mime functions
    * FS#222
    * Keyring -> inline (FireGPGInline)
    * Reorganisation of FireGPG structure in progress.
    * Fixed the about windows

2009-01-26 Maximilien Cuony
    * Fix a problem with form value, thanks to 	Mateusz Baszczyk for his help

2009-01-25 Maximilien Cuony
    * 0.7.2 packed, and uploaded to the server !
    * Version 0.7.2 fix a problem with mail send in plain text (< and > are converted)
    * 0.7.1 packed, and uploaded to the server !
    * Danish added


build.sh
#!/bin/bash -e
# build.sh -- builds JAR and XPI files for mozilla extensions
#   by Nickolay Ponomarev <asqueella@gmail.com>
#   (original version based on Nathan Yergler's build script)
# Most recent version is at <http://kb.mozillazine.org/Bash_build_script>

# This script assumes the following directory structure:
# ./
#   chrome.manifest (optional - for newer extensions)
#   install.rdf
#   (other files listed in $ROOT_FILES)
#
#   content/    |
#   locale/     |} these can be named arbitrary and listed in $CHROME_PROVIDERS
#   skin/       |
#
#   defaults/   |
#   components/ |} these must be listed in $ROOT_DIRS in order to be packaged
#   ...         |
#
# It uses a temporary directory ./build when building; don't use that!
# Script's output is:
# ./$APP_NAME.xpi
# ./$APP_NAME.jar  (only if $KEEP_JAR=1)
# ./files -- the list of packaged files
#
# Note: It modifies chrome.manifest when packaging so that it points to
#       chrome/$APP_NAME.jar!/*

#
# default configuration file is ./config_build.sh, unless another file is
# specified in command-line. Available config variables:
APP_NAME=          # short-name, jar and xpi files name. Must be lowercase with no spaces
CHROME_PROVIDERS=  # which chrome providers we have (space-separated list)
CLEAN_UP=          # delete the jar / "files" when done?       (1/0)
ROOT_FILES=        # put these files in root of xpi (space separated list of leaf filenames)
ROOT_DIRS=         # ...and these directories       (space separated list)
BEFORE_BUILD=      # run this before building       (bash command)
AFTER_BUILD=       # ...and this after the build    (bash command)

if [ -z $1 ]; then
  . ./config_build.sh
else
  . $1
fi

if [ -z $APP_NAME ]; then
  echo "You need to create build config file first!"
  echo "Read comments at the beginning of this script for more info."
  exit;
fi

ROOT_DIR=`pwd`
TMP_DIR=build

#uncomment to debug
#set -x

# remove any left-over files from previous build
rm -f $APP_NAME.jar $APP_NAME.xpi files
rm -rf $TMP_DIR

$BEFORE_BUILD

mkdir --parents --verbose $TMP_DIR/chrome

# generate the JAR file, excluding CVS and temporary files
JAR_FILE=$TMP_DIR/chrome/$APP_NAME.jar
echo "Generating $JAR_FILE..."
for CHROME_SUBDIR in $CHROME_PROVIDERS; do
  find $CHROME_SUBDIR -path '*CVS*' -prune -o -type f -print | grep -v \~ >> files
done

zip -0 -r $JAR_FILE `cat files` -x '*.svn*'
# The following statement should be used instead if you don't wish to use the JAR file
#cp --verbose --parents `cat files` $TMP_DIR/chrome

# prepare components and defaults
echo "Copying various files to $TMP_DIR folder..."
for DIR in $ROOT_DIRS; do
  mkdir $TMP_DIR/$DIR
  FILES="`find $DIR -path '*CVS*' -prune -o -type f -print | grep -v \~`"
  echo $FILES >> files
  cp --verbose --parents $FILES $TMP_DIR
done

# Copy other files to the root of future XPI.
for ROOT_FILE in $ROOT_FILES install.rdf chrome.manifest; do
  cp --verbose $ROOT_FILE $TMP_DIR
  if [ -f $ROOT_FILE ]; then
    echo $ROOT_FILE >> files
  fi
done

cd $TMP_DIR

if [ -f "chrome.manifest" ]; then
  echo "Preprocessing chrome.manifest..."
  # You think this is scary?
  #s/^(content\s+\S*\s+)(\S*\/)$/\1jar:chrome\/$APP_NAME\.jar!\/\2/
  #s/^(skin|locale)(\s+\S*\s+\S*\s+)(.*\/)$/\1\2jar:chrome\/$APP_NAME\.jar!\/\3/
  #
  # Then try this! (Same, but with characters escaped for bash :)
  sed -i -r s/^\(content\\s+\\S*\\s+\)\(\\S*\\/\)$/\\1jar:chrome\\/$APP_NAME\\.jar!\\/\\2/ chrome.manifest
  sed -i -r s/^\(skin\|locale\)\(\\s+\\S*\\s+\\S*\\s+\)\(.*\\/\)$/\\1\\2jar:chrome\\/$APP_NAME\\.jar!\\/\\3/ chrome.manifest

  # (it simply adds jar:chrome/whatever.jar!/ at appropriate positions of chrome.manifest)
fi

# generate the XPI file
echo "Generating $APP_NAME.xpi..."
zip -r ../$APP_NAME.xpi * -x '*.svn*'

cd "$ROOT_DIR"

echo "Cleanup..."
if [ $CLEAN_UP = 0 ]; then
  # save the jar file
  mv $TMP_DIR/chrome/$APP_NAME.jar .
else
  rm ./files
fi

# remove the working files
rm -rf $TMP_DIR
echo "Done!"
$AFTER_BUILD


api_client/firegpgapi.js
/*
 * FireGPG's API version 0.1
 *

***** BEGIN LICENSE BLOCK *****

Version: MPL 1.1/GPL 2.0/LGPL 2.1

The contents of this source code are subject to the Mozilla Public License
Version 1.1 (the "License"); you may not use this source code except in
compliance with the License. You may obtain a copy of the License at
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
for the specific language governing rights and limitations under the
License.

The Original Code is the FireGPG extension.

The Initial Developer of the Original Code is Maximilien Cuony

Portions created by the Initial Developer are Copyright (C) 2007
the Initial Developer. All Rights Reserved.

Contributor(s): Achraf Cherti

Alternatively, the contents of this source code may be used under the terms of
either the GNU General Public License Version 2 or later (the "GPL"), or
the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
in which case the provisions of the GPL or the LGPL are applicable instead
of those above. If you wish to allow use of your version of this source code
only under the terms of either the GPL or the LGPL, and not to allow others to
use your version of this source code under the terms of the MPL, indicate your
decision by deleting the provisions above and replace them with the notice
and other provisions required by the GPL or the LGPL. If you do not delete
the provisions above, a recipient may use your version of this source code
under the terms of any one of the MPL, the GPL or the LGPL.

***** END LICENSE BLOCK *****


 * Website: http://firegpg.com/
 *
 * An API to use FireGPG's functions. You can use the FireGPG's
 * functions directly too, but it's more difficult.
 *
 * ChangeLog :
 *     12/04/08  Achraf Cherti
 *		   - rewrite of test.html, with some enhancements
 *		   - functions "fireGgp<Name>" replaced by "fireGPG<Name>"
 *		   - variables "FireGpg<Name>" replaced by "FireGPGName"
 *		   - The div gpgNode is replaced by firegpg-node
 *
 *     19/01/07  Maximilien Cuony
 *         - The first version of the API, all important
 *           functions are implemented.
 *		   - Version 0.1 released
 */

var FireGPGData;               // Data to send to FireGPG
                               // It's an element in <div id="firegpg-node"></div>

var FireGPGReturnData;         // Data returned by FireGPG
                               // It's an element in <div id="firegpg-node"></div>

var FireGPGHelloOK = false;    // If FireGPG respond
var FireGPGAllowUser = false;  // If the web site is allowed to use FireGPG
var FireGPGAKF = '';           // the AKF is here, after FireGPGAllowUser=true

/*
 * Init the call of FireGPG (it's an internal function)
 */
function fireGPGInitCall() {
	gpgNode = document.getElementById('firegpg-node');
	/* TODO tester si gpgNode est là et faire un return */

	if(FireGPGData == null) {
		FireGPGData = document.createElement("firegpg:data");
		gpgNode.appendChild(FireGPGData);
	}

	if(FireGPGReturnData == null) {
		FireGPGReturnData = document.createElement("firegpg:returndata");
		gpgNode.appendChild(FireGPGReturnData);
	}

	// remove previous data
	attr2 = new Array();
	attrs = FireGPGData.attributes;
	for (i = 0; i < attrs.length; i++)
		attr2[i] = attrs[i]['name'];

	for (i=0; i<attr2.length; i++)
		FireGPGData.removeAttribute(attr2[i]);

	attr2 = new Array();
	attrs = FireGPGReturnData.attributes;
	for (i = 0; i < attrs.length; i++)
		attr2[i] = attrs[i]['name'];

	for (i=0; i<attr2.length; i++)
		FireGPGReturnData.removeAttribute(attr2[i]);
}

/*
 * Check if the FireGPG's API is available.
 *
 * It returns true if FireGPG is available or false.
 */
function fireGPGHello() {
	fireGPGInitCall();
	fireGPGCall('hello');

	if(FireGPGReturnData.hasAttribute('result') &&
	   FireGPGReturnData.getAttribute('result') == 'firegpg-ok') {
		FireGPGHelloOK = true;
		return true;
	}

	return false;
}

/*
 * Authentify your website with an AKF (Authentification Key of FireGPG).
 *
 * On success, true is returned.
 * Else, it's false.
 */
function fireGPGAuth(akf) {
	if(FireGPGHelloOK == false) {
		if(fireGPGHello() == false) {
			/* TODO pas de message dans le fichier firegpgapi.js */
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	fireGPGInitCall();
	FireGPGData.setAttribute('auth_key', akf);
	fireGPGCall('auth');

	if(FireGPGReturnData.hasAttribute('result') &&
	   FireGPGReturnData.getAttribute('result')== 'auth-ok') {
		FireGPGAllowUser = true;
		fireGPGAKF = akf;
		return true;
	}
	else
		return false;
}

/*
 * If the website doesn't have an AKF, this function returns that.
 *
 * On error, false is returned.
 */
function fireGPGRegister() {
	if(FireGPGHelloOK == false) {
		if(fireGPGHello()== false) {
			/* TODO pas de message dans le fichier firegpgapi.js */
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	fireGPGInitCall();
	fireGPGCall('register');

	if(FireGPGReturnData.hasAttribute('result') &&
	   FireGPGReturnData.getAttribute('result')== 'register-ok')
		return FireGPGReturnData.getAttribute('auth_key');
	else
		return false;
}

/*
 * Returne the list of keys (associative key[id] = description).
 *
 * Or false, on error.
 */
function fireGPGListKey(auth_key) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in listprivkey');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);

	fireGPGCall('listkey');

	if(FireGPGReturnData.hasAttribute('result')&& FireGPGReturnData.getAttribute('result')== 'list-ok')
	{
		return_list = new Array();

		list = FireGPGReturnData.getAttribute('list');

		list = list.split(/,/g);

		for (var i=0; i< list.length; i++) {

			key_info = list[i];

			key_info = key_info.split(/:/g);

			if(key_info[0] != undefined && key_info[1] != undefined) {

				return_list[fireGPGAddDoublePoint(key_info[0])] = fireGPGAddDoublePoint(key_info[1]);            }
		}
		return return_list;
	}
	else
	{
		alert('FireGPG - api : Error with listprivkeys');
		return false;
	}

	return;

}

/* TODO : ici */

// Return the list of private key
// Return the keys or false
function fireGPGListPrivKey(auth_key) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in listprivkey');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);

	fireGPGCall('listprivkey');

	if(FireGPGReturnData.hasAttribute('result')&& FireGPGReturnData.getAttribute('result')== 'list-ok')
	{
		return_list = new Array();

		list = FireGPGReturnData.getAttribute('list');

		list = list.split(/,/g);

		for (var i=0; i< list.length; i++) {

			key_info = list[i];

			key_info = key_info.split(/:/g);

			if(key_info[0] != undefined && key_info[1] != undefined) {

				return_list[fireGPGAddDoublePoint(key_info[0])] = fireGPGAddDoublePoint(key_info[1]);            }
		}
		return return_list;
	}
	else
	{
		alert('FireGPG - api : Error with listprivkeys');
		return false;
	}

	return;

}

// Check if a sign is valid.
// Return an object with diffrent informations
function fireGPGCheck(auth_key, text) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in check');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);

	fireGPGCall('check');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in check');
		return false;
	} else {

		var return_object = new Object();

		if(FireGPGReturnData.getAttribute('result')== 'check-ok') {

			return_object.sign_valid = true;
			return_object.infos = FireGPGReturnData.getAttribute('check-infos');


		} else {

			return_object.sign_valid = false;
			return_object.reason = FireGPGReturnData.getAttribute('error'); //unknow, no-gpg, bad-sign, no-key, no-data, bad-pass

		}

		return return_object;
	}

}

// Sign a text
// Return an object with diffrent informations
function fireGPGSign(auth_key, text, forceKey /* Optional */) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in sign');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);

	if(forceKey != undefined)
		FireGPGData.setAttribute('force-key', forceKey);

	fireGPGCall('sign');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in sign');
		return false;
	} else {

		var return_object = new Object();

		if(FireGPGReturnData.getAttribute('result')== 'sign-ok') {

			return_object.sign_ok = true;
			return_object.text = FireGPGReturnData.getAttribute('text');


		} else {

			return_object.sign_ok = false;
			return_object.reason = FireGPGReturnData.getAttribute('error'); //unknow, bad-pass, , user-canceled

		}

		return return_object;
	}

}

// Encrypt a text
// Return an object with diffrent informations
function fireGPGEncrypt(auth_key, text, dest_keys) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in encrypt');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);

	if(dest_keys instanceof Array) {

		dest_keysd = dest_keys[0];


		for (var ii = 1; ii < dest_keys.length; ++ii)
			dest_keysd = dest_keysd + ";" + dest_keys[ii];

		dest_keys = dest_keysd;

	}
	else
		dest_keys = dest_keys + "";


	FireGPGData.setAttribute('keys', dest_keys);

	fireGPGCall('encrypt');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in encrypt');
		return false;
	} else {

		var return_object = new Object();

		if(FireGPGReturnData.getAttribute('result')== 'encrypt-ok') {

			return_object.encrypt_ok = true;
			return_object.text = FireGPGReturnData.getAttribute('text');


		} else {

			return_object.encrypt_ok = false;
			return_object.reason = FireGPGReturnData.getAttribute('error'); //unknow, bad-pass

		}

		return return_object;
	}

}

// Encrypt and sign  a text
// Return an object with diffrent informations
function fireGPGSignAndEncrypt(auth_key, text, dest_keys, forceKey /* Optional */) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in signandencrypt');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);

	if(forceKey != undefined)
		FireGPGData.setAttribute('force-key', forceKey);

	if(dest_keys instanceof Array) {

		dest_keysd = dest_keys[0];


		for (var ii = 1; ii < dest_keys.length; ++ii)
			dest_keysd = dest_keysd + ";" + dest_keys[ii];

		dest_keys = dest_keysd;

	}
	else
		dest_keys = dest_keys + "";


	FireGPGData.setAttribute('keys', dest_keys);

	fireGPGCall('signandencrypt');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in signandencrypt');
		return false;
	} else {

		var return_object = new Object();

		if(FireGPGReturnData.getAttribute('result')== 'signandencrypt-ok') {

			return_object.signandencrypt_ok = true;
			return_object.text = FireGPGReturnData.getAttribute('text');


		} else {

			return_object.signandencrypt_ok = false;
			return_object.reason = FireGPGReturnData.getAttribute('error'); //unknow, bad-pass, , user-canceled

		}

		return return_object;
	}

}

// Decrypt a text
// Return an object with diffrent informations
function fireGPGDecrypt(auth_key, text) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in decrypt');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);


	fireGPGCall('decrypt');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in decrypt');
		return false;
	} else {

		var return_object = new Object();

		if(FireGPGReturnData.getAttribute('result')== 'decrypt-ok') {

			return_object.decrypt_ok = true;


			return_object.text = FireGPGReturnData.getAttribute('text');

			if(FireGPGReturnData.hasAttribute('sign-info'))
				return_object.signInfos = FireGPGReturnData.getAttribute('sign-info');


		} else {

			return_object.decrypt_ok = false;
			return_object.reason = FireGPGReturnData.getAttribute('error'); //unknow, bad-pass, user-canceled

		}

		return return_object;
	}

}

// Check if a sign is valid.
// Return an object with diffrent informations
function fireGPGKimport(auth_key, text) {

	if(FireGPGHelloOK == false)
	{
		if(fireGPGHello()== false) {
			alert('FireGPG - api : Error, FireGPG doesn\'t respond.');
			return false;
		}
	}

	if(FireGPGAllowUser == false)
	{
		if(fireGPGAuth(auth_key)== false) {
			alert('FireGPG - api : Error with auth in check');
			return false;
		}
	}

	fireGPGInitCall();

	FireGPGData.setAttribute('auth_key', auth_key);
	FireGPGData.setAttribute('text', text);

	fireGPGCall('kimport');

	if(FireGPGReturnData.hasAttribute('result')== false)
	{
		alert('FireGPG - api : Error in check');
		return false;
	} else {

		return FireGPGData;
	}

}
// Change #1 to #, #2 to : and #3 to , in a string
function fireGPGAddDoublePoint (s) {

	if(s == undefined)
		return;

	s = s.toString();

	s = s.replace(/#3/g,",");
	s = s.replace(/#2/g,":");
	s = s.replace(/#1/g,"#");

	return s;

}

// Call FireGPG
function fireGPGCall(func ) {

	gpgNode = document.getElementById('firegpg-node');

	var ev = document.createEvent("Events");
	ev.initEvent("firegpg:" + func, true, false);
	gpgNode.dispatchEvent(ev);

}
